iroh/src/node/rpc.rs

```diff
@@ -1037,31 +1020,138 @@ impl<D: BaoStore> Handler<D> {
     }
 }
 
-async fn download_blob<D, C, F>(
-    db: D,
-    get_conn: C,
+async fn download<D>(
+    db: &D,
+    endpoint: MagicEndpoint,
+    downloader: &Downloader,
+    req: BlobDownloadRequest,
+    progress: FlumeProgressSender<DownloadProgress>,
+) -> Result<()>
+where
+    D: iroh_bytes::store::Store,
+{
+    let BlobDownloadRequest {
+        hash,
+        format,
+        nodes,
+        tag,
+        mode,
+    } = req;
+    let hash_and_format = HashAndFormat { hash, format };
+    let stats = match mode {
+        DownloadMode::Queued => {
+            download_queued(
+                endpoint,
+                downloader,
+                hash_and_format,
+                nodes,
+                tag,
+                progress.clone(),
+            )
+            .await?
+        }
+        DownloadMode::Direct => {
+            download_direct_from_nodes(db, endpoint, hash_and_format, nodes, tag, progress.clone())
+                .await?
+        }
+    };
+
+    progress.send(DownloadProgress::AllDone(stats)).await.ok();
+
+    Ok(())
+}
+
+async fn download_queued(
+    endpoint: MagicEndpoint,
+    downloader: &Downloader,
+    hash_and_format: HashAndFormat,
+    nodes: Vec<NodeAddr>,
+    tag: SetTagOption,
+    progress: FlumeProgressSender<DownloadProgress>,
+) -> Result<Stats> {
+    let mut node_ids = Vec::with_capacity(nodes.len());
+    for node in nodes {
+        node_ids.push(node.node_id);
+        endpoint.add_node_addr(node)?;
+    }
+    let req = DownloadRequest::new(hash_and_format, node_ids)
+        .progress_sender(progress)
+        .tag(tag);
+    let handle = downloader.queue(req).await;
+    let stats = handle.await?;
+    Ok(stats)
+}
+
+async fn download_direct_from_nodes<D>(
+    db: &D,
+    endpoint: MagicEndpoint,
     hash_and_format: HashAndFormat,
+    nodes: Vec<NodeAddr>,
     tag: SetTagOption,
-    progress: impl ProgressSender<Msg = DownloadProgress> + IdGenerator,
-) -> Result<()>
+    progress: FlumeProgressSender<DownloadProgress>,
+) -> Result<Stats>
 where
     D: BaoStore,
-    C: FnOnce() -> F,
-    F: Future<Output = Result<quinn::Connection>>,
 {
-    let stats =
-        iroh_bytes::get::db::get_to_db(&db, get_conn, &hash_and_format, progress.clone()).await?;
+    ensure!(!nodes.is_empty(), "No nodes to download from provided.");
+    let mut last_err = None;
+    for node in nodes {
```

Comment:
So this is new functionality introduced in this PR (in #2128 which is now merged into this PR, to be exact). We could parallize this in theory - I mostly added it because it is a common usecase for the downloader, and to keep the RPC APIs for queued vs direct downloads aligned, added this.
* If we keep it serial, the list of nodes is considered ordered, in the order to try the nodes
* We could parallize it, but would then need a good "cutoff point" on when to chose the "winning" request and abort the others (if all succeed). It would also slightly increase the connection overhead, especially if you include many node ids.

So I think this being serial is the simplest for now. I will check that the behavior is documented, and add that "if you want to try connecting to multiple nodes in parallel, use `DownloadMode::Queued`?

Or would you suggest to always parallelize here as well?